lib/sycamore/sycamore/transforms/summarize.py

```diff
@@ -61,133 +73,324 @@ def __init__(self, llm: LLM, element_operator: Optional[Callable[[Element], bool
+        self.data_description = data_description
+        self.max_tokens = max_tokens
+        self.tokenizer = tokenizer
+        self.rounds = 4
+
+    @staticmethod
+    def get_const_vars() -> dict[str, str]:
+        return {
+            "skip_me_key": "_skip_me",
+            "batch_key": "_batch",
+            "round_key": "_round",
+            "intermediate_summary_key": "_summary",
+        }
+
+    def prep_batches(self, doc: Document, round: int = 0) -> Document:
+        vars = self.get_const_vars()
+        for i, elt in enumerate(doc.elements):
+            elt.properties[vars["round_key"]] = round
+            if vars["skip_me_key"] not in elt.properties:
+                elt.properties[vars["skip_me_key"]] = False
+            if elt.properties[vars["skip_me_key"]]:
+                continue
+            this_batch = [i]
+            elt.properties[vars["batch_key"]] = this_batch
+            for j in range(i + 1, len(doc.elements)):
+                e2 = doc.elements[j]
+                if e2.properties.get(vars["skip_me_key"], False):
+                    continue
+                this_batch.append(j)
+                tks = self.prompt.render_element(elt, doc).token_count(self.tokenizer)
+                if tks > self.max_tokens:
+                    this_batch.pop()
+                    break
+                e2.properties[vars["skip_me_key"]] = True
+        return doc
+
+    def cleanup(self, doc: Document) -> Document:
+        if len(doc.elements) == 0:
+            return doc
+        vars = self.get_const_vars()
+        doc.properties["summary"] = doc.elements[0].properties[vars["intermediate_summary_key"]]
+        for e in doc.elements:
+            for v in vars:
+                if v in e.properties:
+                    del e.properties[v]
+        return doc
+
+    def as_llm_map(self, child: Optional[Node], **kwargs) -> Node:
+        vars = self.get_const_vars()
+        if self.fields is not None:
+            self.prompt = self.prompt.set(fields=self.fields)
+        if self.question is not None:
+            self.prompt = self.prompt.set(question=self.question)
+        if self.data_description is not None:
+            self.prompt = self.prompt.set(data_description=self.data_description)
+        nodes = []
+        last = child
+        for round in range(self.rounds):
+            prep_round = Map(child=last, f=self.prep_batches, kwargs={"round": round})
+            llm_round = LLMMapElements(
+                child=prep_round,
+                prompt=self.prompt,
+                output_field=vars["intermediate_summary_key"],
+                llm=self.llm,
+            )
+            nodes.extend([prep_round, llm_round])
+            last = llm_round
+        cleanup = Map(child=last, f=self.cleanup)
+        nodes.append(cleanup)
+        ct = CompositeTransform(child, [])  # type: ignore
```

Comment:
This works but looks weird. I'd do return CompositeTransform(child, nodes=nodes) and fix the constructor.